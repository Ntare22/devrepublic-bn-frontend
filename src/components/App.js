import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import Routes from '../views/routes';
import 'react-toastify/dist/ReactToastify.css';

const App = () => (
  <BrowserRouter>
    <Routes />
  </BrowserRouter>
);
export default App;
