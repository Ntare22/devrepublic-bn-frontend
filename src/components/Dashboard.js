import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import NavBar from './sharedComponents/AuthNavBar';

const Dashboard = ({ token, image }) => (<NavBar token={token} image={image} />);

Dashboard.propTypes = {
  token: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};

const mapStateToProps = ({ user }) => ({
  token: user.token || localStorage.getItem('token'),
  image: '',
});

export default connect(mapStateToProps)(Dashboard);
