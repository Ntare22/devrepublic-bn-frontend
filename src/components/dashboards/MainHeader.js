import React from 'react';
import { Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import translate from '../languages/Translate';
import MainPanel from '../sharedComponents/MainPanel';

const MainHeader = ({ filter, allTrips }) => (
  <Col className="requester-Dashboard">
    <div className="request-dashboard-title ">
      {translate('Manager Dashboard')}
      <MainPanel allTrips={allTrips} filter={filter} />
    </div>
  </Col>
);

MainHeader.propTypes = {
  filter: PropTypes.func,
  allTrips: PropTypes.array,
};

export default MainHeader;
