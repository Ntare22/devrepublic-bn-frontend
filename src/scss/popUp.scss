@import './_variables';
@import '~bootstrap/scss/bootstrap';

.pop-up-container {
    padding: 30px 0;
    position: fixed;
    top: 40%;
    text-align: center;
    left: 50%;
    background: white;
    width: 60vh;
    display: block;
    margin-right: -50%;
    transform: translate(-50%, -50%);
    z-index: 1;
    border: 1px solid;
}

.set-role {
    display: block;
    font-size: 40px;
    font-weight: 500;
    text-align: center;
}

.pop-up-container .form-inline span {
    display: block;
    font-size: 20px;
    padding: 0 40px !important;
    font-weight: 500;
}


.form-inline {
    margin: 0 auto !important
}

.all-users-title {
    font-size: 18px;
    font-weight: 500;
}

.user-dashboard-title {
    font-size: 30px;
    font-weight: 500;
}

.cancel-pop-up, .submit-pop-up {
    background-color: $primary-color;
    border-radius: 0 !important;
    font-size: $normal-text;
    margin-top: 10px;
    padding: 0.5rem 2rem !important;
    &:hover {
      background-color: $secondary-color;
      border-color: $secondary-color;
    }
  }
