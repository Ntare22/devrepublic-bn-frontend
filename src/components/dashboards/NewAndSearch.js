import React from 'react';
import '../../scss/requestTable.scss';
import searchIcon from '../../assets/search-icon.png';
import addIcon from '../../assets/add-icon.png';

const NewAndSearch = () => (
  <div>
    <img className="icon" src={searchIcon} alt="search icon" />
    <img className="icon add-icon" src={addIcon} alt="add icon" />
  </div>
);

export default NewAndSearch;
