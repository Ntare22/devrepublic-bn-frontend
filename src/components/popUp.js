import React, { Component } from 'react';
import PropTypes from 'prop-types';
import translate from './languages/Translate';
import '../scss/popUp.scss';
import Alert from './sharedComponents/Alert';

class PopUp extends Component {
  constructor(props) {
    super(props);
    this.state = {
      roles: this.props.roles,
      email: '',
      role: this.props.activeItem.role,
      errorMessage: '',
      alertVisible: false,
      textColor: '',
    };
  }

  async componentDidMount() {
    await this.setState({ email: '', role: '' });
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      activeItem, isPopUpActive, roles,
    } = nextProps;
    if (activeItem) {
      const email = activeItem.email || '';
      if (nextProps.activeItem.email !== prevState.email) {
        return {
          roles, isPopUpActive, email, role: activeItem.role,
        };
      }
    }
    return null;
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const { email, role } = this.state;
    const response = await this.props.setRoleAction(email, role);
    if (response.payload.status === 200) {
      const users = this.props.users.filter((user) => {
        const returnedUser = user;
        returnedUser.role = returnedUser.email === email ? role : user.role;
        return returnedUser;
      });

      this.props.updateUsers(users);
      this.setState({ alertVisible: false });
      return this.props.handleVisibility('invisible');
    }
    return this.setState({ alertVisible: true, errorMessage: response.payload.data.error.toString(), textColor: 'dark' });
  }

  handleRoleChange = (selected) => {
    const { value } = selected.target;
    this.setState({ role: value });
  }

  render() {
    const title = 'Set Role';
    const cancelButton = translate('Cancel');
    const submitButton = translate('Assign');
    const {
      errorMessage, alertVisible, textColor,
    } = this.state;
    const { activeItem } = this.props;
    return (
      <>
        <div className={`pop-up-container ${this.props.isPopUpActive}`}>
          <div>
            <span className="set-role title">{title}</span>
          </div>
          <br />
          <Alert message={errorMessage} visible={alertVisible} textColor={textColor} otherClassName="pop-up-center" />
          <div className="form-inline">
            <span className="email-text mb-2 mr-sm-5">{translate('User')}</span>
            <h5>
              { activeItem
              && `${activeItem.firstName}  ${activeItem.lastName}`}
            </h5>
          </div>
          <br />
          <div className="form-inline">
            <span className="role-text mb-2 mr-sm-5">{translate('Role')}</span>
            <select
              className="role-field selectpicker show-tick form-control mb-2"
              onChange={this.handleRoleChange}
              name="role"
              placeholder="select user role"
            >
              {this.state.roles.map((role) => {
                const selected = activeItem
                  && role === activeItem.role ? 'selected' : ''
                  || '';
                return (
                  <option value={role} selected={selected}>
                    {role}
                  </option>
                );
              })}
            </select>
          </div>
          <div>
            <button type="button" onClick={() => this.props.handleVisibility('invisible')} className="cancel-pop-up btn btn-danger mb-2 mr-sm-5">{cancelButton}</button>
            <button type="submit" onClick={this.handleSubmit} className="submit-pop-up btn btn-primary mb-2">{submitButton}</button>
          </div>
        </div>
      </>
    );
  }
}

export { PopUp };

PopUp.propTypes = {
  users: PropTypes.array,
  roles: PropTypes.array,
  isPopUpActive: PropTypes.string,
  setRoleAction: PropTypes.func.isRequired,
  handleVisibility: PropTypes.func.isRequired,
  activeItem: PropTypes.object,
  updateUsers: PropTypes.func.isRequired,
};
