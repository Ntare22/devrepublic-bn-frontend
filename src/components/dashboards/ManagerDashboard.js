import React, { Component } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Table from 'react-bootstrap/Table';
import Container from 'react-bootstrap/Container';
import '../../scss/requestTable.scss';
import AuthNavBar from '../sharedComponents/AuthNavBar';
import Footer from '../sharedComponents/Footer';
import { getRequestsAction } from '../../redux/actions/getRequests';
import {
  checkStatus, requestDates, columnTitles, reasonColumnTitle, spinnner, dashboardSidebar,
} from './sharedFuncs';
import NewAndSearch from './NewAndSearch';
import placeholder from '../../assets/users.png';
import translate from '../languages/Translate';
import pagination from '../../util/pagination';
import MainHeader from './MainHeader';


class ManagerDashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      requests: [],
      displayedRequests: [],
      loading: true,
    };
  }

  async componentDidMount() {
    const { token } = this.props;
    const actionOutput = await this.props.getRequestsAction(token);
    this.setState({
      requests: actionOutput.payload.data.requests,
      loading: false,
    });
    this.setState((prevState) => ({ displayedRequests: pagination(prevState.requests, 1) }));
  }

  filter = (newArray) => {
    this.setState({
      displayedRequests: newArray,
    });
  }

  render() {
    const { requests, loading, displayedRequests } = this.state;
    const requestsLength = requests.length;
    const Requester = 'Requester';
    const tripDetails = 'Trip Details';
    const date = 'Date';
    const Status = 'Status';

    return (
      <>
        <AuthNavBar />
        <div>
          <Container className="whole-conatainer" fluid>
            <Row className="bg-dange dashboard-view">
              <Col lg="2" sm="1" xs="1" md="2">
                {dashboardSidebar()}

              </Col>
              <Col lg="10" sm="11" md="9" xs="11" className="col-1">

                <Row>
                  <MainHeader allTrips={requests} filter={this.filter} />
                  <Col className="all-icons text-center">
                    <NewAndSearch />
                  </Col>

                </Row>
                {
                  loading ? (
                    <>
                      {spinnner()}
                    </>
                  ) : ''
                }
                {
                                              requestsLength === 0 && !loading ? (
                                                <div className="no-requests mt-2">
                                                  {translate('No request created by your direct reports yet!')}
                                                </div>
                                              ) : (
                                                <div className="table-header">
                                                  <Table>
                                                    {columnTitles(Requester, tripDetails, date, Status)}
                                                    {displayedRequests.map((request) => (
                                                      <tbody key={request.id}>
                                                        <tr className="row-info-container">
                                                          <td className="col-accomodation">
                                                            <span>
                                                              <img className="profile-image" src={request.User.image || placeholder} alt="profile" />
                                                            </span>
                                                            <span className="firstName">{request.User.firstName}</span>
                                                            <span className="lastName">{request.User.lastName}</span>

                                                          </td>
                                                          {reasonColumnTitle(request)}

                                                          {requestDates(request)}
                                                          <td className="text-center col-status">
                                                            {checkStatus(request)}
                                                          </td>
                                                        </tr>
                                                      </tbody>

                                                    ))}
                                                  </Table>
                                                </div>
                                              )
  }
                <Row />
              </Col>
            </Row>
          </Container>
        </div>
        <Footer />
      </>
    );
  }
}

const MapStateToProps = ({ requests }) => ({
  requests,
});
export { ManagerDashboard };
export default connect(MapStateToProps, { getRequestsAction })(ManagerDashboard);

ManagerDashboard.propTypes = {
  getRequestsAction: PropTypes.func,
  token: PropTypes.string,
};
