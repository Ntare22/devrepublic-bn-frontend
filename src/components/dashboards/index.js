import React, { Component } from 'react';
import queryString from 'query-string';
import PropTypes from 'prop-types';
import jwtDecode from 'jwt-decode';
import RequesterDashboard from './RequesterDashboard';
import ManagerDashboard from './ManagerDashboard';
import AdminDashboard from './admin';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {
      role: '',
      token: '',
    };
  }

  componentDidMount() {
    const { token } = queryString.parse(this.props.location.search);
    const localStorageToken = localStorage.getItem('token');
    const userToken = localStorageToken || token;
    const { role } = jwtDecode(userToken);
    this.setState({ role, token: userToken });
  }

  checkRole = (role) => {
    const { token } = this.state;
    switch (role) {
      case 'requester':
        return <RequesterDashboard token={token} history={this.props.history} />;
      case 'manager':
        return <ManagerDashboard token={token} />;
      case 'super administrator':
        return <AdminDashboard />;
      default:
        return null;
    }
  }

  render() {
    const { role } = this.state;
    return (
      <>
        { this.checkRole(role) }
      </>
    );
  }
}

export default Dashboard;

Dashboard.propTypes = {
  location: PropTypes.object,
  history: PropTypes.object,
};
