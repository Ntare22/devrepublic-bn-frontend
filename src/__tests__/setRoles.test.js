import * as React from 'react';
import { shallow } from 'enzyme';
import dotenv from 'dotenv';
import { act } from 'react-dom/test-utils';
import toJson from 'enzyme-to-json';
import { AdminDashboard } from '../components/dashboards/admin';
import Dashboard from '../components/dashboards/index';
import { PopUp } from '../components/popUp';

dotenv.config();

const user = {
  birthdate: null,
  createdAt: '2020-05-18T13:34:13.981Z',
  currency: null,
  department: null,
  email: 'jean@by.com',
  emailNotifications: true,
  firstName: 'dev',
  gender: null,
  id: '1234',
  image: null,
  isVerified: true,
  language: null,
  lastName: 'devrpo',
  managerId: '12345',
  managerName: 'muhoza',
  oAuthId: 'none',
  residence: null,
  role: 'requester',
  signupMethod: 'none',
  updatedAt: '2020-05-18T13:34:13.981Z',
};

describe('Super Admin dashboard snapshot', () => {
  const dashboardRender = (params, fn = shallow) => {
    const defaultProps = {
      getUsersAction: jest.fn(),
      setRoleAction: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <AdminDashboard {...props} />,
    );
  };
  it('should create a dashboard snapshot', () => {
    expect(toJson(dashboardRender)).toMatchSnapshot();
  });
  it('should call getUsers Action and return users', () => {
    const wrapper = dashboardRender();
    wrapper.setState({
      users: [],
    });
    const { getUsersAction } = wrapper.instance().props;
    getUsersAction.mockReturnValueOnce({
      payload: {
        data: {
          users: [{ id: 'asedr45', email: 'mail@mail.com' }],
        },
      },
    });
    wrapper.setState({
      users: [user],
    });

    expect(wrapper.state().users).toHaveLength(1);
    const button = wrapper.find('.set-role-drop');
    act(() => {
      button.simulate('click');
    });
    wrapper.instance().handleSetRolePopUp('visible', user);
    wrapper.instance().updateUsers([user]);
  });
});

describe('pop Up snapshot', () => {
  const popUpRender = (params, fn = shallow) => {
    const defaultProps = {
      users: [user],
      roles: ['manager', 'travel team member', 'requester', 'travel administrator', 'supplier', 'super administrator'],
      isPopUpActive: '',
      handleVisibility: jest.fn(),
      activeItem: user,
      getUsersAction: jest.fn(),
      setRoleAction: jest.fn(),
      updateUsers: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <PopUp {...props} />,
    );
  };
  it('should create a dashboard snapshot', () => {
    expect(toJson(popUpRender)).toMatchSnapshot();
  });
  it('should submit a form with all info', async () => {
    const popUpWrapper = popUpRender();
    popUpWrapper.setState({
      roles: [],
      email: 'mugisha@gmail.com',
      role: 'requester',
      errorMessage: '',
      alertVisible: false,
      textColor: '',
    });
    const { getUsersAction, setRoleAction } = popUpWrapper.instance().props;
    getUsersAction.mockReturnValueOnce({
      payload: {
        data: {
          users: [{ id: 'asedr45', email: 'mail@mail.com' }],
        },
      },
    });
    setRoleAction.mockReturnValueOnce({
      payload: {
        data: {
          message: 'success',
          status: 200,
        },
      },
    });
    popUpWrapper.setState({
      users: [user],
    });
    const submitButton = popUpWrapper.find('.submit-pop-up');
    const cancelButton = popUpWrapper.find('.cancel-pop-up');
    act(() => {
      const event = { preventDefault: jest.fn() };
      submitButton.simulate('click', event);
      cancelButton.simulate('click');
      popUpWrapper.find('.role-field').simulate('change', { target: { value: 'requester' } });
    });
    expect(submitButton.length).toBe(1);
    expect(cancelButton.length).toBe(1);
  });
});

describe('Index dashboard snapshot', () => {
  const render = (params, fn = shallow) => {
    const defaultProps = {
      getUsersActionn: jest.fn(),
      setRolesActionn: jest.fn(),
      location: {},
      search: jest.fn(),
    };
    const props = { ...defaultProps, ...params };
    return fn(
      <Dashboard {...props} />,
    );
  };
  beforeEach(() => {
    localStorage.setItem('token', process.env.ADMIN_TOKEN);
  });
  it('should create a dashboard container snapshot', () => {
    expect(toJson(render)).toMatchSnapshot();
  });
  it('check the user status before rendering dashboard', () => {
    const wrapper = render();
    wrapper.setState({
      role: '',
    });
    wrapper.setState({
      role: 'super administrator',
    });
    expect(wrapper.state().role).toEqual('super administrator');
    wrapper.instance().checkRole({ role: 'super administrator' });
  });
});
