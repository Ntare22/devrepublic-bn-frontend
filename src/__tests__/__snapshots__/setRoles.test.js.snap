// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index dashboard snapshot should create a dashboard container snapshot 1`] = `undefined`;

exports[`Super Admin dashboard snapshot should create a dashboard snapshot 1`] = `undefined`;

exports[`pop Up snapshot should create a dashboard snapshot 1`] = `undefined`;
